{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "from glob import glob\n",
    "\n",
    "import yaml\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import patches\n",
    "from matplotlib import ticker\n",
    "import seaborn\n",
    "import numpy\n",
    "import pandas as pd\n",
    "import networkx\n",
    "import dinopy\n",
    "import pybedtools\n",
    "\n",
    "from sh import bwa, samtools\n",
    "\n",
    "from phasm.io import gfa\n",
    "from phasm.assembly_graph import AssemblyGraph\n",
    "from phasm.bubbles import find_superbubbles\n",
    "\n",
    "BASE_DIR = os.path.realpath(os.path.join(os.getcwd(), '..'))\n",
    "ALIGNED_CONTIGS_FILE = os.path.join(\n",
    "    BASE_DIR, \"assemblies/{assembly}/05_analysis/aligned_contigs.bam\")\n",
    "\n",
    "with open(os.path.join(BASE_DIR, \"config.yml\")) as f:\n",
    "    config = yaml.load(f)\n",
    "    \n",
    "#seaborn.set_context('paper')\n",
    "seaborn.set_style('whitegrid')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align contigs to reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating bam file...\n",
      "\n",
      "Creating bam file...\n",
      "\n",
      "Creating bam file...\n",
      "\n",
      "Creating bam file...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for assembly, asm_config in config['assemblies'].items():\n",
    "    parts = assembly.split('-')\n",
    "    ploidy = int(parts[0].replace(\"ploidy\", \"\"))\n",
    "    coverage = int(parts[1].replace(\"x\", \"\"))\n",
    "    \n",
    "    if coverage != 60:\n",
    "        continue\n",
    "    \n",
    "    asm_dir = os.path.join(BASE_DIR, \"assemblies\", assembly)\n",
    "    ref_fasta = os.path.join(BASE_DIR, asm_config['reference'])\n",
    "    bam_file = ALIGNED_CONTIGS_FILE.format(assembly=assembly)\n",
    "    contigs_fasta = os.path.join(asm_dir, \"04_phase\", assembly + \".fasta\")\n",
    "    \n",
    "    os.makedirs(os.path.dirname(bam_file), exist_ok=True)\n",
    "    \n",
    "    if not os.path.isfile(ref_fasta + '.bwt'):\n",
    "        print(\"Create reference index...\")\n",
    "        # No BWA index available, create one\n",
    "        bwa.index(ref_fasta)\n",
    "    \n",
    "    if not os.path.isfile(bam_file) and os.path.isfile(contigs_fasta):\n",
    "        print(\"Creating bam file...\")\n",
    "        samtools.sort(samtools.view(bwa.mem(ref_fasta, contigs_fasta), '-uS', '-'), '-o', bam_file)\n",
    "        \n",
    "    if not os.path.isfile(bam_file + '.bai') and os.path.isfile(bam_file):\n",
    "        samtools.index(bam_file)\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average_contig_len</th>\n",
       "      <th>contig_fraction</th>\n",
       "      <th>coverage</th>\n",
       "      <th>haploblock_fraction</th>\n",
       "      <th>longest_contig</th>\n",
       "      <th>ng50</th>\n",
       "      <th>num_contigs</th>\n",
       "      <th>num_contigs_norm</th>\n",
       "      <th>ploidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>128312.863636</td>\n",
       "      <td>0.181175</td>\n",
       "      <td>60</td>\n",
       "      <td>0.818825</td>\n",
       "      <td>282556</td>\n",
       "      <td>39952</td>\n",
       "      <td>22</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>88329.217391</td>\n",
       "      <td>0.078573</td>\n",
       "      <td>60</td>\n",
       "      <td>0.921427</td>\n",
       "      <td>195203</td>\n",
       "      <td>195197</td>\n",
       "      <td>46</td>\n",
       "      <td>15.333333</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>130623.625000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>60</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>328364</td>\n",
       "      <td>122790</td>\n",
       "      <td>48</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>195952.229167</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>60</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>307809</td>\n",
       "      <td>209731</td>\n",
       "      <td>48</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   average_contig_len  contig_fraction  coverage  haploblock_fraction  \\\n",
       "0       128312.863636         0.181175        60             0.818825   \n",
       "1        88329.217391         0.078573        60             0.921427   \n",
       "2       130623.625000         0.000000        60             1.000000   \n",
       "3       195952.229167         0.000000        60             1.000000   \n",
       "\n",
       "   longest_contig    ng50  num_contigs  num_contigs_norm  ploidy  \n",
       "0          282556   39952           22         11.000000       2  \n",
       "1          195203  195197           46         15.333333       3  \n",
       "2          328364  122790           48         12.000000       4  \n",
       "3          307809  209731           48          8.000000       6  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "for assembly, asm_config in config['assemblies'].items():\n",
    "    parts = assembly.split('-')\n",
    "    ploidy = int(parts[0].replace(\"ploidy\", \"\"))\n",
    "    coverage = int(parts[1].replace(\"x\", \"\"))\n",
    "    \n",
    "    asm_dir = os.path.join(BASE_DIR, \"assemblies\", assembly)\n",
    "    ref_fasta = os.path.join(BASE_DIR, asm_config['reference'])\n",
    "    bam_file = ALIGNED_CONTIGS_FILE.format(assembly=assembly)\n",
    "    contigs_fasta = os.path.join(asm_dir, \"04_phase\", assembly + \".fasta\")\n",
    "    \n",
    "    if coverage != 60:\n",
    "        continue\n",
    "    \n",
    "    if not os.path.isfile(contigs_fasta):\n",
    "        continue\n",
    "    \n",
    "    ref_reader = dinopy.FastaReader(ref_fasta)\n",
    "    ref_length = sum(e.length for e in ref_reader.entries())\n",
    "    \n",
    "    theo_ng50 = ref_length / 2\n",
    "    ng50 = 0\n",
    "    \n",
    "    contig_reader = dinopy.FastaReader(contigs_fasta)\n",
    "    total_contig_length = 0\n",
    "    total_haploblock_length = 0\n",
    "    contig_sequence_len = 0\n",
    "    contig_lengths = []\n",
    "    for contig in contig_reader.entries():\n",
    "        total_contig_length += contig.length\n",
    "        \n",
    "        if theo_ng50 <= total_contig_length:\n",
    "            ng50 = contig.length\n",
    "            \n",
    "        name = contig.name.decode('utf-8')\n",
    "        if \"haploblock\" in name:\n",
    "            total_haploblock_length += contig.length\n",
    "        else:\n",
    "            contig_sequence_len += contig.length\n",
    "        \n",
    "        contig_lengths.append(contig.length)\n",
    "        \n",
    "    data.append({\n",
    "        'ploidy': ploidy,\n",
    "        'coverage': coverage,\n",
    "        'num_contigs': len(contig_lengths),\n",
    "        'num_contigs_norm': len(contig_lengths) / ploidy,\n",
    "        'ng50': ng50,\n",
    "        'longest_contig': max(contig_lengths),\n",
    "        'average_contig_len': sum(contig_lengths) / len(contig_lengths),\n",
    "        'haploblock_fraction': total_haploblock_length / total_contig_length,\n",
    "        'contig_fraction': contig_sequence_len / total_contig_length\n",
    "    })\n",
    "        \n",
    "df = pd.DataFrame(data)\n",
    "df "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Percentage of reference genome covered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def contig_filter(f, name):\n",
    "    return f[0] == name\n",
    "\n",
    "for assembly, asm_config in config['assemblies'].items():\n",
    "    parts = assembly.split('-')\n",
    "    ploidy = int(parts[0].replace(\"ploidy\", \"\"))\n",
    "    coverage = int(parts[1].replace(\"x\", \"\"))\n",
    "    \n",
    "    asm_dir = os.path.join(BASE_DIR, \"assemblies\", assembly)\n",
    "    ref_fasta = os.path.join(BASE_DIR, asm_config['reference'])\n",
    "    bam_file = ALIGNED_CONTIGS_FILE.format(assembly=assembly)\n",
    "    contigs_fasta = os.path.join(asm_dir, \"04_phase\", assembly + \".fasta\")\n",
    "    \n",
    "    if not os.path.isfile(contigs_fasta):\n",
    "        continue\n",
    "        \n",
    "    if coverage != 60:\n",
    "        continue\n",
    "    \n",
    "    top_margin = 0.4  # inches\n",
    "    bottom_margin = 0.6  # inches\n",
    "    total_margin = top_margin + bottom_margin\n",
    "    axis_height = 0.8\n",
    "    hspace = 0.2\n",
    "    all_axis_height = (ploidy*axis_height) + ((ploidy-1) * (axis_height*hspace))\n",
    "    \n",
    "    margin_fraction = total_margin / (all_axis_height + total_margin)\n",
    "    top_margin_frac = (top_margin * margin_fraction) / total_margin\n",
    "    bottom_margin_frac = (bottom_margin * margin_fraction) / total_margin\n",
    "    \n",
    "    fig_height = all_axis_height + total_margin\n",
    "    fig, ax = plt.subplots(ploidy, sharex=True, figsize=(7, fig_height))\n",
    "    bt = pybedtools.BedTool()\n",
    "    fr = dinopy.FastaReader(ref_fasta)\n",
    "    for i, homolog in enumerate(fr.entries()):\n",
    "        name = homolog.name.decode('utf-8')\n",
    "        chr_id = name.split()[0]\n",
    "        \n",
    "        cov = bt.genome_coverage(ibam=bam_file, bga=True, stream=True).filter(\n",
    "            contig_filter, chr_id)\n",
    "        \n",
    "        coverage_data = numpy.zeros((homolog.length,))\n",
    "        for region in cov:\n",
    "            start = int(region[1])\n",
    "            end = int(region[2])\n",
    "            reg_coverage = int(region[3])\n",
    "            \n",
    "            coverage_data[start:end] = reg_coverage\n",
    "        \n",
    "        ax[i].plot(coverage_data)\n",
    "        ax[i].set_xlim((0, 800000))\n",
    "        ax[i].set_ylim((0, 15))\n",
    "        ax[i].set_ylabel(\"Chr. {}\".format(i+1))\n",
    "        ax[i].yaxis.set_major_locator(ticker.MultipleLocator(5))\n",
    "        \n",
    "        ax[i].add_patch(\n",
    "            patches.Rectangle((0, 0), homolog.length, 0.5, alpha=0.3)\n",
    "        )\n",
    "        \n",
    "    fig.subplots_adjust(bottom=bottom_margin_frac,\n",
    "                        top=(1-top_margin_frac), hspace=hspace)\n",
    "    fig.text(0.5, (1-top_margin_frac)+(top_margin_frac/2), \"Coverage by PHASM contigs (genome: ploidy {}, {}x read coverage)\".format(ploidy, coverage),\n",
    "             ha=\"center\")\n",
    "    fig.text(0.5, bottom_margin_frac-((bottom_margin_frac/2) + (bottom_margin_frac/4)), \"Chromosome position [bases]\",\n",
    "             ha=\"center\")\n",
    "    fig.text(0.025, 0.5, \"Coverage\", rotation=\"vertical\", va=\"center\")\n",
    "    fig.savefig(os.path.join(BASE_DIR, 'figures', assembly + '-cov.png'), dpi=256, transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  },
  "widgets": {
   "state": {},
   "version": "1.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
